src/finn/custom_op/fpgadataflow/thresholding_binary_search.py

```diff
@@ -0,0 +1,752 @@
+            )
+
+        # Create an RTL block, not an IP core (-type ip)
+        cmd.append(
+            "create_bd_cell -type module -reference %s %s"
+            % (self.get_nodeattr("gen_top_module"), self.onnx_node.name)
+        )
+
+        # ERROR: [BD 41-237] Bus Interface property FREQ_HZ does not match between
+        # /Thresholding_Binary_Search_0/s_axis(100000000 and
+        # /StreamingFIFO_0/out_V(200000000.000000)
+        cmd.append(
+            "set_property -dict [list CONFIG.FREQ_HZ {200000000}] [%s %s]"
+            % ("get_bd_intf_pins", "Thresholding_Binary_Search_0/s_axis")
+        )
+
+        # ERROR: [BD 41-237] Bus Interface property FREQ_HZ does not match between
+        # /StreamingFIFO_1/in0_V(200000000.000000) and
+        # /Thresholding_Binary_Search_0/m_axis(100000000)
+        cmd.append(
+            "set_property -dict [list CONFIG.FREQ_HZ {200000000}] [%s %s]"
+            % ("get_bd_intf_pins", "Thresholding_Binary_Search_0/m_axis")
+        )
+
+        return cmd
+
+    def get_verilog_top_module_intf_names(self):
+        """Return a dict of names of input and output interfaces.
+        The keys reflect the protocols each interface implements:
+        'clk', 'rst', 'm_axis', 's_axis', 'aximm', 'axilite'.
+        Values are lists of tuples (axis, aximm) or names (axilite):
+        'axis' tuples correspond to the list of node inputs in order,
+        each tuple is (interface_name, interface_width_bits).
+        axilite always assumed to be 32 bits and is not tuple (name only).
+        Each block must have at most one aximm and one axilite."""
+
+        intf_names = super().get_verilog_top_module_intf_names()
+        # Only 'decoupled' mode is supported - check before adding axilite interface
+        mem_mode = self.get_nodeattr("mem_mode")
+        if mem_mode != "decoupled":
+            raise Exception(
+                "Unrecognized memory mode for this node: {}".format(mem_mode)
+            )
+        intf_names["axilite"] = ["s_axilite"]
+        intf_names["s_axis"] = [["s_axis"]]
+        intf_names["m_axis"] = [["m_axis"]]
+
+        self.set_nodeattr("runtime_writeable_weights", 1)
+
+        return intf_names
+
+    def find_next_power_of_2(self, n):
+        # Negative values will loop infinitely below - return 0
+        if n <= 0:
+            return 0
+        # If '1' is requested, output will be '0' in the loop below, avoid this now.
+        elif n == 1:
+            return 2  # i.e. 2**1
+
+        # decrement 'n' (to handle cases when `n` itself is a power of 2)
+        n = n - 1
+
+        # loop until only one bit is left
+        while n & n - 1:
+            # unset rightmost bit
+            n = n & n - 1
+        return n << 1
+
+    def twos_comp(self, val, bitwidth):
+        return (val + (1 << bitwidth)) % (1 << bitwidth)
+
+    def prep_axilite_val(self, val):
+        return self.twos_comp(int(val), self.get_weight_datatype().bitwidth())
```

Comment:
there's a possibility here to use existing functions from `data_packing.py` like `pack_innermost_dim_as_hex_string`-- e.g. see how the decoupled_verilog_dat generation mode operates in `make_weight_file`

Reply:
done: https://github.com/Xilinx/finn/pull/687/commits/ca0042225c006d4545e26b0e0f1221ecd4ab68c3